### PR TITLE
.vcdlist implementation

### LINKED ISSUE #160: Unhandled vcdlist_c in Dota 2
What is it?

### PR DESCRIPTION
**Pull request by krotter0**
Still needs a bunch of work and cleaning up, but it can handle all HLA .vcdlists already.

Fixes #160

### CODE REVIEW

File: ValveResourceFormat/Resource/ResourceTypes/Choreo/Parser/BVCDParser.cs [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+        {
+            this.reader = reader;
+            this.strings = strings;
+        }
+
+        public static ChoreoData Parse(Stream stream, string[] strings)
+        {
+            using BinaryReader reader = new BinaryReader(stream);
+
+            var magic = new string(reader.ReadChars(4));
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/Choreo/Parser/BVCDParser.cs:30]: Read as uint, add MAGIC const field (we do this for all other formats)

File: ValveResourceFormat/Resource/ResourceTypes/Choreo/Parser/BVCDParser.cs [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+            }
+            var version = reader.ReadByte();
+            var crc = reader.ReadInt32();
+
+            BVCDParser parser;
+            switch (version)
+            {
+                case 8: //There are some v8 bvcds in HLA, but they decompile fine with the v9 parser.
+                case 9:
+                    parser = new BVCDParserV9(reader, strings);
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/Choreo/Parser/BVCDParser.cs:43]: I'd rather not create new classes for each version. Usually version bumps just add/modify fields and its one `if` away.

File: ValveResourceFormat/Resource/ResourceTypes/Choreo/Parser/BVCDParser.cs [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+                    break;
+                case 17:
+                    parser = new BVCDParserV17(reader, strings);
+                    break;
+                default:
+                    throw new NotImplementedException($"BVCD version {version} is not supported");
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/Choreo/Parser/BVCDParser.cs:49]: Use UnexpectedMagicException

File: ValveResourceFormat/Resource/ResourceTypes/Choreo/Data/ChoreoEvent.cs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        public bool UsingRelativeTag { get; private set; }
+        public ChoreoRelativeTag RelativeTag { get; private set; }
+        public ChoreoEventFlex EventFlex { get; private set; }
+        public byte LoopCount { get; private set; }
+        public ChoreoClosedCaptions ClosedCaptions { get; private set; }
+        public int Id { get; private set; }
+        public int Unk01 { get; private set; }
+
+        //todo: ew
+        public ChoreoEvent(ChoreoEventType type, string name, float startTime, float endTime, string param1, string param2, string param3, ChoreoRamp ramp, ChoreoFlags flags, float distanceToTarget, ChoreoRelativeTag[] relativeTags, ChoreoFlexTimingTag[] flexTimingTags, ChoreoAbsoluteTag[] absoluteTags, float sequenceDuration, bool usingRelativeTag, ChoreoRelativeTag relativeTag, ChoreoEventFlex eventFlex, byte loopCount, ChoreoClosedCaptions closedCaptions, int id, int unk01)
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/Choreo/Data/ChoreoEvent.cs:31]: You can change `private set` to `init` and set them as attributes straight away.

File: ValveResourceFormat/Resource/ResourceTypes/Choreo/Data/ChoreoSample.cs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            outKV.AddProperty("deg", new KVValue(KVType.FLOAT, Bezier.Value.OutDegrees));
+            outKV.AddProperty("weight", new KVValue(KVType.FLOAT, Bezier.Value.OutWeight));
+            kv.AddProperty("out", new KVValue(KVType.OBJECT, outKV));
+
+            return kv;
+        }
+
+        private static string GetCurveTypeName(byte index)
+        {
+            switch (index)
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/Choreo/Data/ChoreoSample.cs:95]: Use pattern matching switch

File: ValveResourceFormat/Resource/ResourceTypes/Choreo/Data/ChoreoSample.cs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+                case 0x07:
+                    return "kochanek";
+                case 0x08:
+                    return "kochanek_early";
+                case 0x09:
+                    return "kochanek_late";
+                case 0x10:
+                    return "bezier";
+                default:
+                    throw new NotImplementedException();
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/Choreo/Data/ChoreoSample.cs:120]: Use UnexpectedMagicException

File: ValveResourceFormat/Resource/ResourceTypes/Choreo/Parser/BVCDParser.cs [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+                    var inType = reader.ReadByte();
+                    var outType = reader.ReadByte();
+                    var unk01 = reader.ReadByte(); //what's this
+                    Debug.Assert(unk01 == 0); //Does this have to be 0?
+
+                    lastSample.SetCurveType(inType, outType);
+                }
+                else
+                {
+                    throw new NotImplementedException($"Unexpected choreo sample data type ({type})");
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/Choreo/Parser/BVCDParser.cs:272]: Use UnexpectedMagicException

File: ValveResourceFormat/Resource/ResourceTypes/Choreo/Data/ChoreoEvent.cs [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+                //case ChoreoEventType.iklockleftarm: return "iklockleftarm";
+                //case ChoreoEventType.iklockrightarm: return "iklockrightarm";
+                case ChoreoEventType.SubScene: return "subscene";
+                case ChoreoEventType.Interrupt: return "interrupt";
+                case ChoreoEventType.PermitResponses: return "permitresponses";
+                case ChoreoEventType.Camera: return "camera";
+                case ChoreoEventType.Loop: return "loop";
+                case ChoreoEventType.Section: return "section";
+                case ChoreoEventType.StopPoint: return "stoppoint"; //TODO: verify stoppoint event's name
+                default: throw new NotImplementedException();
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/Choreo/Data/ChoreoEvent.cs:149]: Use UnexpectedMagicException and pattern matching switch

File: ValveResourceFormat/Resource/ResourceTypes/Choreo/Data/ChoreoFlexSample.cs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+            "kochanek",
+            "kochanek_early",
+            "kochanek_late",
+            "simple_cubic",
+            "catmullrom",
+            "catmullrom_normalize",
+            "catmullrom_tangent",
+            "exponential_decay",
+            "hold"
+        ];
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/Choreo/Data/ChoreoFlexSample.cs:22]: There's also at least "bezier" after "hold"

File: ValveResourceFormat/Resource/ResourceTypes/Choreo/Data/ChoreoEvent.cs [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+            kv.AddProperty("end_time", new KVValue(KVType.FLOAT, EndTime));
+            kv.AddProperty("param", new KVValue(KVType.STRING, Param1));
+            kv.AddProperty("param2", new KVValue(KVType.STRING, Param2));
+            kv.AddProperty("param3", new KVValue(KVType.STRING, Param3));
+
+            if (ClosedCaptions != null)
+            {
+                var ccType = ClosedCaptions.Type switch {
+                    ChoreoClosedCaptionsType.Master => "cc_master",
+                    ChoreoClosedCaptionsType.Slave => "cc_slave", //TODO: is this actually named cc_slave?
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/Choreo/Data/ChoreoEvent.cs:74]: Seems like it. There's also cc_disabled

File: ValveResourceFormat/Resource/ResourceTypes/Choreo/Parser/BVCDParser.cs [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+            var eventId = reader.ReadInt32();
+
+            var absoluteTags = playTags.Concat(shiftTags).ToArray();
+            return new ChoreoEvent(eventType, name, eventStart, eventEnd, param1, param2, param3, ramp, flags, distanceToTarget, relativeTags, flexTimingTags, absoluteTags, sequenceDuration, usingRelativeTag, relativeTag, flex, loopCount, closedCaptions, eventId, unk01);
+        }
+        protected virtual ChoreoClosedCaptions ReadClosedCaptions()
+        {
+            var type = (ChoreoClosedCaptionsType)reader.ReadByte();
+            var token = ReadString();
+            var flags = (ChoreoClosedCaptionsFlags)reader.ReadByte();
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/Choreo/Parser/BVCDParser.cs:241]: There's also SoundStartDelay float after these flags.

There's also flag 8, but i'm unsure what it means.

File: ValveResourceFormat/Resource/ResourceTypes/Choreo/Parser/BVCDParser.cs [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+                var unk02 = reader.ReadInt32(); //is this for speak only? what's this? maybe part of closedcaptions?
+                Debug.Assert(unk02 == 0);
+            }
+
+            if (version >= 17)
+            {
+                var unk03 = reader.ReadByte();
+            }
+
+            //eventId or unk01 is sometimes missing?
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/Choreo/Parser/BVCDParser.cs:229]: After this line it doesn't seem like these bytes belong to choreo event

File: ValveResourceFormat/Resource/ResourceTypes/Choreo/ChoreoData.cs [hunk 13]
```diff
@@ -0,0 +1,60 @@
+using System.Linq;
+using ValveResourceFormat.ResourceTypes.Choreo.Data;
+using ValveResourceFormat.Serialization.KeyValues;
+
+namespace ValveResourceFormat.ResourceTypes.Choreo
+{
+    public class ChoreoData
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/Choreo/ChoreoData.cs:7]: ChoreoScene

File: ValveResourceFormat/Resource/ResourceTypes/Choreo/Parser/BVCDParser.cs (hunk 12 above)
Comment (krotter0) [reply ValveResourceFormat/Resource/ResourceTypes/Choreo/Parser/BVCDParser.cs]: They're definitely there, but I can't see how they get there either.
The server has <16 vcd version.

**Review by kristiker — APPROVED** [review]

File: ValveResourceFormat/Resource/Enums/ResourceType.cs [hunk 14]
```diff
@@ -81,7 +81,7 @@ public enum ResourceType
         ResourceRemapTable,
 
         [Extension("vcdlist")]
-        ChoreoSceneFileData,
+        ChoreoDataList,
```
Comment (xPaw) [inline ValveResourceFormat/Resource/Enums/ResourceType.cs:84]: The identifier was correct before: "CompileChoreoSceneFileData"

File: ValveResourceFormat/Resource/ResourceTypes/ChoreoDataList.cs [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+            scene.SoundDuration = sceneSoundDuration;
+            scene.HasSounds = hasSounds != 0;
+
+            return scene;
+        }
+
+        private static byte[] ReadSceneBlock(BinaryReader reader, int length)
+        {
+            var id = new string(reader.ReadChars(4));
+            if (id == "LZMA")
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/ChoreoDataList.cs:82]: Read as uint

File: ValveResourceFormat/Resource/ResourceTypes/ChoreoDataList.cs [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+            {
+                var uncompressedLength = reader.ReadInt32();
+                var compressedLength = reader.ReadInt32();
+
+                var lzmaDecoder = new LzmaDecoder();
+                lzmaDecoder.SetDecoderProperties(reader.ReadBytes(5));
+                var compressedBuffer = reader.ReadBytes(compressedLength);
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/ChoreoDataList.cs:89]: Use ArrayPool

Comment (krotter0) [reply ValveResourceFormat/Resource/ResourceTypes/ChoreoDataList.cs]: I've removed the compressedBuffer variable entirely, I'm guessing that's where you meant it. I was considering returning an arraypool array too, but that'd probably be a bit awkward to handle for functions that call this.

Comment (xPaw) [reply ValveResourceFormat/Resource/ResourceTypes/ChoreoDataList.cs]: Yep i meant compressed buffer, using the stream directly is also good.

**Comment by xPaw** [comment]
Good to merge?

**Comment by krotter0** [comment]
Should be. I was able to decompile all vcdlists without errors, and recompile them in engine too.

**Pull request merged**